docs/decisions/0021-json-serializable-custom-types.md

```diff
@@ -0,0 +1,163 @@
+---
+status: proposed
+contact: dehoward
+date: 2023-11-06
+deciders: alliscode, markwallace-microsoft
+consulted:
+informed:
+---
+
+# JSON Serializable Custom Types
+
+## Context and Problem Statement
+
+This ADR aims to simplify the usage of custom types by allowing developers to use any type that can be serialized using `System.Text.Json`.
+
+Standardizing on a JSON-serializable type is necessary to allow functions to be described using a JSON Schema within a planner's function manual. Using a JSON Schema to describe a function's input and output types will allow the planner to validate that the function is being used correctly.
+
+Today, use of custom types within Semantic Kernel requires developers to implement a custom `TypeConverter` to convert to/from the string representation of the type. This is demonstrated in [Example60_AdvancedNativeFunctions](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/KernelSyntaxExamples/Example60_AdvancedNativeFunctions.cs#L202C44-L202C44) as seen below:
+
+```csharp
+    [TypeConverter(typeof(MyCustomTypeConverter))]
+    private sealed class MyCustomType
+    {
+        public int Number { get; set; }
+
+        public string? Text { get; set; }
+    }
+
+    private sealed class MyCustomTypeConverter : TypeConverter
+    {
+        public override bool CanConvertFrom(ITypeDescriptorContext? context, Type sourceType) => true;
+
+        public override object? ConvertFrom(ITypeDescriptorContext? context, CultureInfo? culture, object value)
+        {
+            return JsonSerializer.Deserialize<MyCustomType>((string)value);
+        }
+
+        public override object? ConvertTo(ITypeDescriptorContext? context, CultureInfo? culture, object? value, Type destinationType)
+        {
+            return JsonSerializer.Serialize(value);
+        }
+    }
+```
+
+## Considered Options
+
+**1. Use TypeConverter for primitive types and native serialization methods for custom types**
```

Comment:
Can the summary be improved, is the following accurate

1. Primitive types will be handled using JSON serialization
2. Complex types will be handled by the registered `TypeConverter` is available
3. If no `TypeConverter` is registered for a complex type, JSON serialization/deserialization is used
    - This option may fail in which case a clear error message is provided explaining a  `TypeConverter` is required

Reply:
yes, this is accurate. I've updated the desription to better reflect this.